### PR TITLE
Pushbullet pull request unification

### PR DESCRIPTION
**Pull request by dsander**
This supersedes #635, #636 and #640

I also added a migration for older `PushbulletAgent`s which only supported the `note` message type.

@berendt Thanks for the commits this is based on!

### CODE REVIEW

File: app/models/agents/pushbullet_agent.rb [hunk 1]
```diff
@@ -41,19 +60,22 @@ def working?
...
     end
 
     private
-
     def query_options(event)
       mo = interpolated(event)
       {
-        :basic_auth => {:username => mo[:api_key], :password => ''},
-        :body => {:device_iden => mo[:device_id], :title => mo[:title], :body => mo[:body], :type => 'note'}
+        :basic_auth => {username: mo[:api_key], password: ''},
```
Comment (cantino) [inline app/models/agents/pushbullet_agent.rb]: (I'm definitely a fan of moving toward the 1.9-style hashes.)

**Comment by cantino** [comment]
Awesome!

**Comment by dsander** [comment]
Ok, I am done with that one for now. Another review would be welcome :smile:

File: app/models/agents/pushbullet_agent.rb [hunk 2]
```diff
@@ -25,14 +37,39 @@ def default_options
...
+    form_configurable :body, type: :text
+    form_configurable :url
+    form_configurable :name
+    form_configurable :address
+
     def validate_options
-      errors.add(:base, "you need to specify a pushbullet api_key") unless options['api_key'].present?
-      errors.add(:base, "you need to specify a device_id") if options['device_id'].blank?
+      errors.add(:base, "you need to specify a pushbullet api_key") if options['api_key'].blank?
+      create_device if options['device_id'].blank?
```
Comment (cantino) [inline app/models/agents/pushbullet_agent.rb]: Bit strange to make changes during validation.  Any other way to do this, or does it at least avoid making more than one?

File: db/migrate/20150219213604_add_type_option_attribute_to_pushbullet_agents.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+class AddTypeOptionAttributeToPushbulletAgents < ActiveRecord::Migration
+  def up
+    Agents::PushbulletAgent.all.each do |agent|
```
Comment (cantino) [inline db/migrate/20150219213604_add_type_option_attribute_to_pushbullet_agents.rb]: I doubt users have thousands of these, but `find_each` is generally safer.

File: app/models/agents/pushbullet_agent.rb (hunk 2 above)
Comment (dsander) [inline app/models/agents/pushbullet_agent.rb]: You are right, I could move it to a `before_validation` callback, would that make more sense?

Comment (cantino) [inline app/models/agents/pushbullet_agent.rb]: In general, it seems like it should be a before_save, so validations don't have side effects?

**Comment by berendt** [comment]
Thanks for taking over. Please close the superseded pull request after the merge of this PR.

File: app/models/agents/pushbullet_agent.rb (hunk 2 above)
Comment (dsander) [inline app/models/agents/pushbullet_agent.rb]: With a before_save the agent could end up being saved without having a `device_id` assigned if the `create_device` call fails. 
I am mostly using before_validation callsback to set up model attributes about which the user does not need to take about (in this case if the `create_device` method succeeds).

Comment (cantino) [inline app/models/agents/pushbullet_agent.rb]: I'll leave this up to you.  I think either will be okay.

**Pull request merged**